processor/groupbytraceprocessor/processor.go

```diff
@@ -0,0 +1,344 @@
+//
+// The typical data flow looks like:
+// ConsumeTraces() -> chan received -> onTrace() -> processBatch() -> processTrace() ->
+// time.AfterFunc(chan expired) -> onTrace() -> processExpired() -> async markAsReleased() ->
+// chan released -> onReleased() -> nextConsumer.ConsumeTraces()
+type groupByTraceProcessor struct {
+	nextConsumer consumer.TraceConsumer
+	config       Config
+	logger       *zap.Logger
+
+	// used for the ring buffer
+	index     int
+	ids       [][]byte
+	idToIndex map[uint64]int // key is hash(traceID), value is the index on the 'ids' slice
+
+	// the trace storage
+	st storage
+
+	// traces received from the previous processors
+	received chan pdata.Traces
+
+	// traceID to be released
+	expired chan []byte
+
+	// released traces
+	released chan *v1.ResourceSpans
+
+	// traceID to be removed
+	removed chan []byte
+
+	// shutdown synchronization
+	stopped        chan context.Context // used in the main loop
+	cancelReleases bool                 // used for traces that are just waiting for the duration
+	cancelLock     *sync.RWMutex
+}
+
+var _ component.TraceProcessor = (*groupByTraceProcessor)(nil)
+
+// newGroupByTraceProcessor returns a new processor.
+func newGroupByTraceProcessor(logger *zap.Logger, st storage, nextConsumer consumer.TraceConsumer, config Config) (*groupByTraceProcessor, error) {
+	if nextConsumer == nil {
+		return nil, componenterror.ErrNilNextConsumer
+	}
+
+	sp := &groupByTraceProcessor{
+		logger:       logger.With(zap.String("processor", "groupbytrace")),
+		nextConsumer: nextConsumer,
+		config:       config,
+		st:           st,
+		ids:          make([][]byte, config.NumTraces),
+		idToIndex:    make(map[uint64]int),
+		index:        -1, // the first span to be received will be placed at position '0'
+
+		// open the channels
+		removed:  make(chan []byte, 50), // TODO: perf tests to decide what's a good buffer size
+		expired:  make(chan []byte, 50),
+		received: make(chan pdata.Traces, 50),
+		released: make(chan *v1.ResourceSpans, 50),
+
+		// shutdown synchronization
+		stopped:        make(chan context.Context),
+		cancelLock:     &sync.RWMutex{},
+		cancelReleases: false,
+	}
+
+	return sp, nil
+}
+
+func (sp *groupByTraceProcessor) ConsumeTraces(_ context.Context, td pdata.Traces) error {
+	sp.received <- td
+	return nil
+}
+
+func (sp *groupByTraceProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (sp *groupByTraceProcessor) Start(_ context.Context, _ component.Host) error {
+	// starts the processes listening to the channels
+	go sp.onReleased()
+	go sp.onRemoved()
+	go sp.onTrace()
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (sp *groupByTraceProcessor) Shutdown(ctx context.Context) error {
+	sp.stopped <- ctx
```

Comment:
Do you flush all buffered traces on Shutdown?

Reply:
Most of the operations are executed as part of the main loop, so, the effect should be minimal in this implementation. When real on disk storage is implemented, the disk write might be async, so, a shutdown signal will need to be propagated down. After the shutdown call is received, the traces waiting for the duration will be discarded once their wait duration elapses.

Reply:
> Most of the operations are executed as part of the main loop

You forward traces to the nextConsumer asynchronously via `go sp.markAsReleased(traceID)` after `wait_duration` is expired, which is not part of the main loop.

> After the shutdown call is received, the traces waiting for the duration will be discarded once their wait duration elapses.

The default `wait_duration` is 1 second and can be configured to be even bigger.

@tigrannajaryan @bogdandrutu what is expected graceful shutdown behavior? I would assume that we should flush all buffers to export any remaining data rather than just dropping them.

Reply:
IMO, those traces should be discarded -- we have been asked explicitly by the user to wait for a specific duration before dispatching the trace. Sending it before will potentially dispatch incomplete traces, breaking one of the use cases for this processor.

Reply:
> the trace may not complete within expected duration, so consumer must be tolerant to incomplete traces IMO.

I think we have different expectations here: when using this processor, the intention of our users is to receive complete traces, by waiting a specific amount of time before dispatching it. My expectation is that incomplete traces would be discarded, your expectation is that incomplete traces should be passed down the pipeline. Should we make a flag to control this? I can see that users will be confused either way if we don't document this well and provide a way to override the default behavior.